File: ingestion/better_raw/llm.py
"""Shared NVIDIA (Nemotron) client, OpenAI-compatible.

Swapped from Anthropic to NVIDIA's build.nvidia.com endpoint for the Nemotron
hackathon track. Structured output and tool calling follow the OpenAI Chat
Completions contract; reasoning depth is set via extra_body rather than a
`thinking` param, and there's no hosted web-search tool — the search backend is
wired up separately in alerts.py against DuckDuckGo.
"""

from __future__ import annotations

from functools import lru_cache

import openai

from .config import settings

NVIDIA_BASE_URL = "https://integrate.api.nvidia.com/v1"


@lru_cache(maxsize=1)
def client() -> openai.OpenAI:
    return openai.OpenAI(api_key=settings().nvidia_api_key, base_url=NVIDIA_BASE_URL)


def model() -> str:
    return settings().nvidia_model  # e.g. "nvidia/nemotron-3-super-120b-a12b"


# Nemotron's reasoning-depth knob — the analogue of Anthropic's `thinking`.
# Use it where the model has a judgement to make.
REASONING = {
    "chat_template_kwargs": {"enable_thinking": True},
    "reasoning_budget": 8192,
}

# Reasoning tokens come out of the same output budget as the answer, so turn it
# off for jobs that are pure formatting.
NO_REASONING = {
    "chat_template_kwargs": {"enable_thinking": False},
}

File: src/lib/llm.ts
import "server-only";

import OpenAI from "openai";

import { requireEnv } from "@/lib/env";

/**
 * Nemotron over NVIDIA's OpenAI-compatible endpoint — the same model and key
 * the Python ingestion worker uses. There is no Anthropic key in the
 * environment any more, so nothing here may fall back to one.
 */

const NVIDIA_BASE_URL = "https://integrate.api.nvidia.com/v1";

let cached: OpenAI | null = null;

export function llm(): OpenAI {
  if (!cached) {
    cached = new OpenAI({
      apiKey: requireEnv("NVIDIA_API_KEY"),
      baseURL: NVIDIA_BASE_URL,
    });
  }
  return cached;
}

export function llmModel(): string {
  return process.env.NVIDIA_MODEL ?? "nvidia/nemotron-3-super-120b-a12b";
}

/**
 * Thinking off. The alerts already contain the reasoning — this call only has
 * to read them and answer in the buyer's terms. Reasoning tokens would come
 * out of the same output budget while adding several seconds to what should
 * feel like a chat response.
 *
 * NVIDIA reads this as a top-level request field, which is not in the OpenAI
 * schema, so callers spread it into the params object and cast. The SDK
 * serialises whatever it is given.
 */
export const NO_REASONING = {
  chat_template_kwargs: { enable_thinking: false },
};